Mathlib/Data/Tree.lean

```diff
@@ -0,0 +1,186 @@
+TODO: implement `traversable tree`. -/
+def map {β} (f : α → β) : Tree α → Tree β
+  | nil => nil
+  | node a l r => node (f a) (map f l) (map f r)
+#align tree.map Tree.map
+
+/-- The number of internal nodes (i.e. not including leaves) of a binary tree -/
+@[simp]
+def numNodes : Tree α → ℕ
+  | nil => 0
+  | node _ a b => a.numNodes + b.numNodes + 1
+#align tree.num_nodes Tree.numNodes
+
+/-- The number of leaves of a binary tree -/
+@[simp]
+def numLeaves : Tree α → ℕ
+  | nil => 1
+  | node _ a b => a.numLeaves + b.numLeaves
+#align tree.num_leaves Tree.numLeaves
+
+/-- The height - length of the longest path from the root - of a binary tree -/
+@[simp]
+def height : Tree α → ℕ
+  | nil => 0
+  | node _ a b => max a.height b.height + 1
+#align tree.height Tree.height
+
+theorem numLeaves_eq_numNodes_succ (x : Tree α) : x.numLeaves = x.numNodes + 1 := by
+  induction x <;> simp [*, Nat.add_comm, Nat.add_assoc, Nat.add_left_comm]
+#align tree.num_leaves_eq_num_nodes_succ Tree.numLeaves_eq_numNodes_succ
+
+theorem numLeaves_pos (x : Tree α) : 0 < x.numLeaves := by
+  rw [numLeaves_eq_numNodes_succ]
+  exact x.numNodes.zero_lt_succ
+#align tree.num_leaves_pos Tree.numLeaves_pos
+
+theorem height_le_numNodes : ∀ x : Tree α, x.height ≤ x.numNodes
+  | nil => le_rfl
+  | node _ a b =>
+    Nat.succ_le_succ
+      (max_le (_root_.trans a.height_le_numNodes <| a.numNodes.le_add_right _)
+        (_root_.trans b.height_le_numNodes <| b.numNodes.le_add_left _))
+#align tree.height_le_num_nodes Tree.height_le_numNodes
+
+/-- The left child of the tree, or `nil` if the tree is `nil` -/
+@[simp]
+def left : Tree α → Tree α
+  | nil => nil
+  | node _ l _r => l
+#align tree.left Tree.left
+
+/-- The right child of the tree, or `nil` if the tree is `nil` -/
+@[simp]
+def right : Tree α → Tree α
+  | nil => nil
+  | node _ _l r => r
+#align tree.right Tree.right
+
+-- mathport name: «expr △ »
+-- Notation for making a node with `Unit` data
+scoped infixr:65 " △ " => Tree.node ()
+
+-- porting note: workaround for leanprover/lean4#2049
+section recursor_workarounds
+
+/-- A computable version of `Tree.recOn`. Workaround until Lean has native support for this. -/
+def recOnC {α} {motive : Tree α → Sort u} (t : Tree α) (base : motive Tree.nil)
+  (ind : (a : α) → (l : Tree α) → (r : Tree α) → motive l → motive r → motive (Tree.node a l r))
+  : motive t :=
+  match t with
+  | nil => base
+  | Tree.node a l r => ind a l r (recOnC l base ind) (recOnC r base ind)
+
+@[csimp]
+lemma recOn_Unit_eq_recOnC : @Tree.recOn = @Tree.recOnC := by
+  ext α motive t base ind
+  induction t with
+  | nil => rfl
+  | node a l r ihl ihr =>
+    rw [Tree.recOnC, ←ihl, ←ihr]
+
+end recursor_workarounds
+
+@[elab_as_elim]
+def unitRecOn {motive : Tree Unit → Sort _} (t : Tree Unit) (base : motive nil)
+    (ind : ∀ x y, motive x → motive y → motive (x △ y)) : motive t :=
+    t.recOn base fun _ => ind
```

Comment:
Can you add a porting note saying you removed `Unit.recOn` as structure eta makes it unnecessary?

Reply:
Done.